### ISSUE TITLE
Incorrect HDF5 Version

### ISSUE THREAD

**Issue by pgierz** [conda-forge/cdo-feedstock#62]
Issue:
When install `cdo` with Conda on MacOS, the `hdf5` library is downgraded to 1.10.4, but cdo wants 1.10.5.

```
conda install -c conda-forge cdo                                     ✔  INSERT  (LIG)      10150  09:30:16  
Collecting package metadata (current_repodata.json): done
Solving environment: failed with current_repodata.json, will retry with next repodata source.
Initial quick solve with frozen env failed.  Unfreezing env and trying again.
Solving environment: failed with current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed
Initial quick solve with frozen env failed.  Unfreezing env and trying again.
Solving environment: done

## Package Plan ##

  environment location: /Users/pgierz/anaconda3/envs/LIG

  added / updated specs:
    - cdo


The following NEW packages will be INSTALLED:

  bzip2              conda-forge/osx-64::bzip2-1.0.8-h01d97ff_0
  cdo                conda-forge/osx-64::cdo-1.9.7.1-h561daef_0
  curl               conda-forge/osx-64::curl-7.63.0-heae2a1f_1000
  eccodes            conda-forge/osx-64::eccodes-2.12.3-hac8aea4_1
  fftw               conda-forge/osx-64::fftw-3.3.8-mpi_openmpi_h2385e66_1006
  hdf4               conda-forge/osx-64::hdf4-4.2.13-hf3c6af0_1002
  jasper             conda-forge/osx-64::jasper-1.900.1-h636a363_1006
  krb5               conda-forge/osx-64::krb5-1.16.2-h24a3359_1000
  libaec             conda-forge/osx-64::libaec-1.0.4-h0a44026_0
  libcurl            conda-forge/osx-64::libcurl-7.63.0-h76de61e_1000
  libnetcdf          conda-forge/osx-64::libnetcdf-4.6.2-h6b88ef6_1001
  libssh2            conda-forge/osx-64::libssh2-1.8.0-hf30b1f0_1003
  libxml2            conda-forge/osx-64::libxml2-2.9.9-hd80cff7_2
  llvm-openmp        conda-forge/osx-64::llvm-openmp-8.0.0-h770b8ee_0
  mpi                conda-forge/osx-64::mpi-1.0-openmpi
  openmpi            conda-forge/osx-64::openmpi-4.0.1-ha90c164_1
  ossuuid            conda-forge/osx-64::ossuuid-1.6.2-h0a44026_1000
  udunits2           conda-forge/osx-64::udunits2-2.2.27.6-h776b7f1_1001

The following packages will be DOWNGRADED:

  hdf5                           1.10.5-nompi_h0cbb7df_1100 --> 1.10.4-nompi_h0cbb7df_1106

```

Although the install works normally, when trying to use cdo, the following error is produced:
```
$ cdo --version
dyld: Library not loaded: @rpath/libhdf5.103.dylib
  Referenced from: /Users/pgierz/anaconda3/envs/LIG/bin/cdo
  Reason: Incompatible library version: cdo requires version 105.0.0 or later, but libhdf5.103.dylib provides version 104.0.0
[2]    3472 abort      cdo --version
```

I guess this is connected to an incorrect dependency being pinned, but I don't know where to change that...

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /Users/pgierz/anaconda3/envs/LIG:
#
# Name                    Version                   Build  Channel
appnope                   0.1.0                    py37_0    conda-forge
asn1crypto                0.24.0                py37_1003    conda-forge
attrs                     19.1.0                   py37_1  
backcall                  0.1.0                    py37_0  
bcrypt                    3.1.6            py37h01d97ff_1    conda-forge
bleach                    3.1.0                    py37_0  
bzip2                     1.0.8                h01d97ff_0    conda-forge
ca-certificates           2019.6.16            hecc5488_0    conda-forge
cartopy                   0.17.0          py37hdb866aa_1005    conda-forge
cdo                       1.5.3                    pypi_0    pypi
certifi                   2019.6.16                py37_1    conda-forge
cffi                      1.12.3           py37hccf1714_0    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
cryptography              2.5              py37hdbc3d79_1    conda-forge
curl                      7.63.0            heae2a1f_1000    conda-forge
cycler                    0.10.0                     py_1    conda-forge
dbus                      1.13.6               h90a0687_0  
decorator                 4.4.0                    py37_1  
defusedxml                0.6.0                      py_0  
eccodes                   2.12.3               hac8aea4_1    conda-forge
entrypoints               0.3                      py37_0  
expat                     2.2.6                h0a44026_0  
fftw                      3.3.8           mpi_openmpi_h2385e66_1006    conda-forge
freetype                  2.10.0               h24853df_0    conda-forge
geos                      3.7.2                h6de7cb9_1    conda-forge
gettext                   0.19.8.1             h15daf44_3  
glib                      2.56.2               hd9629dc_0  
hdf4                      4.2.13            hf3c6af0_1002    conda-forge
hdf5                      1.10.4          nompi_h0cbb7df_1106    conda-forge
icu                       58.2                 h4b95b61_1  
idna                      2.8                   py37_1000    conda-forge
ipykernel                 5.1.1            py37h39e3cac_0  
ipython                   7.6.1            py37h39e3cac_0  
ipython_genutils          0.2.0                    py37_0  
ipywidgets                7.4.2                    py37_0  
jasper                    1.900.1           h636a363_1006    conda-forge
jedi                      0.13.3                   py37_0    conda-forge
jinja2                    2.10.1                   py37_0  
jpeg                      9c                h1de35cc_1001    conda-forge
json5                     0.8.5                      py_0    conda-forge
jsonschema                3.0.1                    py37_0    conda-forge
jupyter                   1.0.0                    py37_7  
jupyter_client            5.2.4                    py37_0  
jupyter_console           6.0.0                    py37_0  
jupyter_core              4.5.0                      py_0  
jupyterlab                1.0.2                    py37_0    conda-forge
jupyterlab_server         1.0.0                      py_1    conda-forge
kiwisolver                1.1.0            py37h770b8ee_0    conda-forge
krb5                      1.16.2            h24a3359_1000    conda-forge
libaec                    1.0.4                h0a44026_0    conda-forge
libblas                   3.8.0               10_openblas    conda-forge
libcblas                  3.8.0               10_openblas    conda-forge
libcurl                   7.63.0            h76de61e_1000    conda-forge
libcxx                    4.0.1                hcfea43d_1  
libcxxabi                 4.0.1                hcfea43d_1  
libedit                   3.1.20181209         hb402a30_0  
libffi                    3.2.1                h475c297_4  
libgfortran               3.0.1                         0    conda-forge
libiconv                  1.15                 hdd342a3_7  
liblapack                 3.8.0               10_openblas    conda-forge
libnetcdf                 4.6.2             h6b88ef6_1001    conda-forge
libopenblas               0.3.6                hd44dcd8_4    conda-forge
libpng                    1.6.37               h2573ce8_0    conda-forge
libsodium                 1.0.16               h3efe00b_0  
libssh2                   1.8.0             hf30b1f0_1003    conda-forge
libtiff                   4.0.10            hd08fb8f_1003    conda-forge
libxml2                   2.9.9                hd80cff7_2    conda-forge
llvm-openmp               8.0.0                h770b8ee_0    conda-forge
lz4-c                     1.8.3             h6de7cb9_1001    conda-forge
markupsafe                1.1.1            py37h1de35cc_0    conda-forge
matplotlib                3.1.1                    py37_0    conda-forge
matplotlib-base           3.1.1            py37h3a684a6_0    conda-forge
mistune                   0.8.4            py37h1de35cc_0  
mpi                       1.0                     openmpi    conda-forge
nbconvert                 5.5.0                      py_0    conda-forge
nbformat                  4.4.0                    py37_0  
ncurses                   6.1                  h0a44026_1  
notebook                  5.7.8                    py37_0    conda-forge
numpy                     1.16.4           py37h6b0580a_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openblas                  0.3.6                hd44dcd8_4    conda-forge
openmpi                   4.0.1                ha90c164_1    conda-forge
openssl                   1.0.2r               h1de35cc_0    conda-forge
ossuuid                   1.6.2             h0a44026_1000    conda-forge
owslib                    0.18.0                     py_0    conda-forge
pandas                    0.24.2           py37h86efe34_0    conda-forge
pandoc                    2.2.3.2                       0  
pandocfilters             1.4.2                    py37_1  
paramiko                  2.6.0                    py37_0    conda-forge
parso                     0.5.0                      py_0    conda-forge
pcre                      8.43                 h0a44026_0  
pexpect                   4.7.0                    py37_0    conda-forge
pickleshare               0.7.5                    py37_0  
pillow                    5.3.0           py37hbddbef0_1000    conda-forge
pip                       19.1.1                   py37_0    conda-forge
proj4                     6.1.0                h2cc77ee_2    conda-forge
prometheus_client         0.7.1                      py_0    conda-forge
prompt_toolkit            2.0.9                    py37_0  
ptyprocess                0.6.0                    py37_0    conda-forge
pycparser                 2.19                     py37_1    conda-forge
pyepsg                    0.4.0                      py_0    conda-forge
pygments                  2.4.2                      py_0    conda-forge
pykdtree                  1.3.1           py37h917ab60_1002    conda-forge
pynacl                    1.3.0           py37h1de35cc_1000    conda-forge
pyopenssl                 19.0.0                   py37_0    conda-forge
pyparsing                 2.4.0                      py_0    conda-forge
pyproj                    2.2.1            py37h804dea5_0    conda-forge
pyqt                      5.9.2            py37h655552a_2  
pyrsistent                0.14.11          py37h1de35cc_0    conda-forge
pyshp                     2.1.0                      py_0    conda-forge
pysocks                   1.7.0                    py37_0    conda-forge
python                    3.7.0                hc167b69_0  
python-dateutil           2.8.0                      py_0    conda-forge
pytz                      2019.1                     py_0    conda-forge
pyzmq                     18.0.0           py37h0a44026_0  
qt                        5.9.7                h468cd18_1  
qtconsole                 4.5.1                      py_0    conda-forge
readline                  7.0                  h1de35cc_5  
requests                  2.22.0                   py37_0    conda-forge
scipy                     1.3.0            py37hab3da7d_0    conda-forge
send2trash                1.5.0                    py37_0  
setuptools                41.0.1                   py37_0    conda-forge
shapely                   1.6.4           py37h0567c5e_1006    conda-forge
sip                       4.19.8           py37h0a44026_0  
six                       1.12.0                py37_1000    conda-forge
sqlite                    3.28.0               ha441bb4_0  
terminado                 0.8.2                    py37_0    conda-forge
testpath                  0.4.2                    py37_0  
tk                        8.6.8                ha441bb4_0    conda-forge
tornado                   6.0.3            py37h01d97ff_0    conda-forge
tqdm                      4.32.2                     py_0    conda-forge
traitlets                 4.3.2                    py37_0    conda-forge
udunits2                  2.2.27.6          h776b7f1_1001    conda-forge
urllib3                   1.24.3                   py37_0    conda-forge
wcwidth                   0.1.7                    py37_0  
webencodings              0.5.1                    py37_1  
wheel                     0.33.4                   py37_0    conda-forge
widgetsnbextension        3.4.2                    py37_0  
xarray                    0.12.3                     py_0    conda-forge
xlrd                      1.2.0                      py_0    conda-forge
xz                        5.2.4                h1de35cc_4  
zeromq                    4.3.1                h0a44026_3  
zlib                      1.2.11               h1de35cc_3  
zstd                      1.4.0                ha9f0a20_0    conda-forge

```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
     active environment : LIG
    active env location : /Users/pgierz/anaconda3/envs/LIG
            shell level : 2
       user config file : /Users/pgierz/.condarc
 populated config files : /Users/pgierz/.condarc
          conda version : 4.7.10
    conda-build version : 3.18.5
         python version : 3.7.3.final.0
       virtual packages : 
       base environment : /Users/pgierz/anaconda3  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/osx-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/pgierz/anaconda3/pkgs
                          /Users/pgierz/.conda/pkgs
       envs directories : /Users/pgierz/anaconda3/envs
                          /Users/pgierz/.conda/envs
               platform : osx-64
             user-agent : conda/4.7.10 requests/2.21.0 CPython/3.7.3 Darwin/18.2.0 OSX/10.14.2
                UID:GID : 952471260:355342394
             netrc file : None
           offline mode : False

```
</details>

**Comment by pgierz** [comment]
I've been trying to install the dependencies one by one so I can then use the `--no-deps` option. (I know, this is dangerous). This lead me to figure out that it seems as if this actually doesn't come from the CDO feedstock, but rather from the dependency `libnetcdf`.

**Comment by valeriupredoi** [comment]
hey @pgierz we are running in a somewhat related issue see https://github.com/ESMValGroup/ESMValTool/issues/1239 - the problem w are faced with is that in an environment that needs to have deps built against `mpi` with `mpich` the new cdo would't install (via conda) since its deps are set to build against `mpi` with `openmpi`; you will be happy to hear that a barebones environment with only `hdf5-1.10.5` and `cdo=1.9.7.1` gets built correctly on Linux x64 and cdo works well, so it sounds to me that your issue is OSX-specific

**Comment by pgierz** [comment]
> it sounds to me that your issue is OSX-specific 🍺

Super... 🤦‍♂️ 
Another entry in my long list of arguments to finally get around to ditching Apple once and for all. 

I've worked around this for now by installing all the requirements one by one and passing `--no-deps` to prevent anything from downgrading the `hdf5` again. While that does "solve the problem", it's not particularly elegant.

For anyone who stumbles across this in future. If you get something like this:
```
$ conda install hdf5
# Now, try to install something with dependencies:
$ conda install cdo

...conda does some stuff...

The following NEW packages will be INSTALLED:
A
B
C
D

The following packages will be DOWNGRADED:

  hdf5 
```

stop the installation, and go through `A`, `B`, `C`, `D` by hand, each time checking if it wants to modify hdf5, and if so, installing any subsequent dependencies until it **just** wants to install `A`, and then pass `--no-deps`. You'll get warnings about an inconsistent environment, and this will with *high probability* cause major problems at some point, but, it might be a temporary solution...

**Comment by Try2Code** [comment]
From the CDO perspective, MPI is **not needed at all**. It's absolutely safe to drop it. 

Maybe the dependencies of the cdo package can be changed to leave out mpi completely @ocefpaf

**Comment by Try2Code** [comment]
@pgierz if possible, you could use macports for cdo as a (maybe temporary) replacement. just as a fall-back

**Comment by valeriupredoi** [comment]
> From the CDO perspective, MPI is **not needed at all**. It's absolutely safe to drop it.
> 
> Maybe the dependencies of the cdo package can be changed to leave out mpi completely @ocefpaf

Excellent! In that case maybe you guys can edit the `info/repodata_record.json` file in the conda package to drop the `mpi_openmpi` against `fftw` since that bugger is messing our environment solvability - it installs:
```
fftw                      3.3.8           mpi_openmpi_h6dd7431_1006    conda-forge
mpi                       1.0                     openmpi    conda-forge
```
which are not loved at all by our `mpich`-based dependencies :grin:

**Comment by ocefpaf** [comment]
In theory the build should not pull the `mpi` alternative unless `mpich` or `openmpi` is present. I'll take a look at hit tomorrow.

**Comment by ocefpaf** [comment]
I cannot reproduce the problem locally. Looks like @pgierz environment is mixing packages from `defaults` and `conda-forge`. See 
https://conda-forge.org/docs/user/tipsandtricks.html#how-to-fix-it for instructions on how to use `conda-forge`.

**Comment by valeriupredoi** [comment]
@ocefpaf any news on making `cdo` dependencies mpi-agnostic? Cheers muchly :beer:

**Comment by ocefpaf** [comment]
> @ocefpaf any news on making `cdo` dependencies mpi-agnostic? Cheers muchly

What do you get when installing it in a clean env by following https://conda-forge.org/docs/user/tipsandtricks.html#how-to-fix-it and with:

```
conda create --name TEST cdo
```

**Comment by ocefpaf** [comment]
I trimmed to only the packages of interest and here is the problem:

```yaml
    host:
        - fftw 3.3.8 nompi_h5c49c53_1105
        - hdf5 1.10.5 nompi_h0cbb7df_1100
    run:
        - fftw * mpi_openmpi_*
```

While `fftw` without `mpi` is used at build time, somehow conda-build used the `openmpi` in `run`. I believe this is a bug with conda-build and/or how we are dealing with the `mpi` strings.

**Comment by valeriupredoi** [comment]
Environment file: `cdoenv.yml`:
```
---
name: cdoenv
channels:
  - conda-forge

dependencies:
  - cdo=1.9.7.1
```

Workflow:
```
conda env create -n cdoenv -f cdoenv.yml
conda activate cdoenv
```

Result:
```
(cdoenv) valeriu@valeriu-PORTEGE-Z30-C:~$ conda list mpi
# packages in environment at /home/valeriu/anaconda3R/envs/cdoenv:
#
# Name                    Version                   Build  Channel
mpi                       1.0                     openmpi    conda-forge
openmpi                   4.0.1                hc99cbb1_1    conda-forge
```

Conclusion: bummer! :grin:

So you trimmed the dep from:
```
  "depends": [
    "eccodes",
    "fftw * mpi_openmpi_*",
```

and conda-build still builds it and outputs the `repodata_record` with `openmpi`?

**Comment by ocefpaf** [comment]
Can you try what I mentioned in https://github.com/conda-forge/cdo-feedstock/issues/62#issuecomment-520860433 and send the full `conda list`?

Note that we must activate `strict` as it is specified in the instructions above!

**Comment by valeriupredoi** [comment]
I can't set the `channel_priority` variable in `.condarc` because I don't have a `.condarc` and nonetheless we don't want to start telling our users to start creating extra files when they don't need to - our package uses multiple channels and those should be listed in the `.condarc` and this is prone to user errors. In any case, for test purposes, I created a `.condarc` and populated it with:
```
channel_priority: strict
channels:
  - conda-forge
  - defaults
```

then
```
conda create --name TEST cdo
conda activate TEST
conda list

# packages in environment at /home/valeriu/anaconda3R/envs/TEST:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
bzip2                     1.0.8                h516909a_0    conda-forge
ca-certificates           2019.6.16            hecc5488_0    conda-forge
cdo                       1.9.7.1              ha44cbb3_0    conda-forge
curl                      7.65.3               hf8cf82a_0    conda-forge
eccodes                   2.13.1               h83cb724_0    conda-forge
expat                     2.2.5             he1b5a44_1003    conda-forge
fftw                      3.3.8           mpi_openmpi_h6dd7431_1006    conda-forge
hdf4                      4.2.13            h9a582f1_1002    conda-forge
hdf5                      1.10.5          nompi_h3c11f04_1100    conda-forge
icu                       64.2                 he1b5a44_0    conda-forge
jasper                    1.900.1           h07fcdf6_1006    conda-forge
jpeg                      9c                h14c3975_1001    conda-forge
krb5                      1.16.3            h05b26f9_1001    conda-forge
libaec                    1.0.4                hf484d3e_0    conda-forge
libcurl                   7.65.3               hda55be3_0    conda-forge
libedit                   3.1.20170329      hf8c457e_1001    conda-forge
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libiconv                  1.15              h516909a_1005    conda-forge
libnetcdf                 4.6.2             h056eaf5_1002    conda-forge
libpng                    1.6.37               hed695b0_0    conda-forge
libssh2                   1.8.2                h22169c7_2    conda-forge
libstdcxx-ng              9.1.0                hdf63c60_0  
libxml2                   2.9.9                hee79883_2    conda-forge
mpi                       1.0                     openmpi    conda-forge
ncurses                   6.1               hf484d3e_1002    conda-forge
openmpi                   4.0.1                hc99cbb1_1    conda-forge
openssl                   1.1.1c               h516909a_0    conda-forge
ossuuid                   1.6.2             hf484d3e_1000    conda-forge
proj4                     6.1.0                he751ad9_2    conda-forge
readline                  8.0                  hf8c457e_0    conda-forge
sqlite                    3.29.0               hcee41ef_0    conda-forge
tk                        8.6.9             hed695b0_1002    conda-forge
udunits2                  2.2.27.6          h4e0c4b3_1001    conda-forge
xz                        5.2.4             h14c3975_1001    conda-forge
zlib                      1.2.11            h516909a_1005    conda-forge
```

**Comment by ocefpaf** [comment]
> I can't set the `channel_priority` variable in `.condarc` because I don't have a `.condarc`

Not sure that is true, conda always creates a `.condarc`. What is the result of conda info?

Also, you can can set with a command:

```
conda config --set channel_priority strict
```

> and nonetheless we don't want to start telling our users to start creating extra files when they don't need to 

It is required to get support from `conda-forge`, otherwise it is impossible for us to debug and you are on your own. Also, `strict` is recommended when using multiple channels and will be default in future versions of `conda`.

> our package uses multiple channels and those should be listed in the `.condarc` and this is prone to user errors.

If you use more channel you have more reasons to use `strict`.

In any case, for test purposes, I created a `.condarc` and populated it with.

Yep, now, as you can see, you got the right `hdf5`. The `fftw` problem is what I mentioned in https://github.com/conda-forge/cdo-feedstock/issues/62#issuecomment-520862623 and should be fixed in future conda-build (if that is not fixed already.).

**Comment by valeriupredoi** [comment]
`hdf5` versioning has never been a problem for us - the problem is the new `cdo` that asks for `openmpi` via the `fftw` dependency (whereas we need to use `mpi` with `mpich`) or, if downgrading `cdo` to `1.9.6` the problem becomes the (erroneous?) message that `cdo` was built against the wrong `hdf5` version (ie 1.10.5, even if `cdo -V` says `1.10.4(1.10.5)`). Can you please comment on the first problem (the openmpi one) - when will it be fixed *from your cdo side* (apologies if you are not a `cdo` dev) :beer:

**Comment by ocefpaf** [comment]
> `hdf5` versioning has never been a problem for us - the problem is the new `cdo` that asks for `openmpi` via the `fftw` dependency (whereas we need to use `mpi` with `mpich`) or, if downgrading `cdo` to `1.9.6` the problem becomes the (erroneous?) message that `cdo` was built against the wrong `hdf5` version (ie 1.10.5, even if `cdo -V` says `1.10.4(1.10.5)`). Can you please comment on the first problem (the openmpi one)

`hdf5` was the original problem in @pgierz issue. Not sure what you mean there.

> when will it be fixed _from your cdo side_ (apologies if you are not a `cdo` dev)


There is nothing to fix with `nco`. See my comment above about the conda build bug and the `fftw` in `run`.

TL;DR you must use `strict` to the the right set of packages and avoid the mix in @pgierz original issue.

Your issue is a bug in conda-build that, once solved, will propagate to a new `cdo` package.

**Comment by valeriupredoi** [comment]
> Your issue is a bug in conda-build that, once solved, will propagate to a new cdo package.

Excellent news! Can you pls give an estimate for the new `cdo` package? Cheers :beer:

Note: posting on this issue is probably misleading for both you and the OP since our issue (see https://github.com/ESMValGroup/ESMValTool/issues/1239 is totally different than the OP's :grin: )

**Comment by Try2Code** [comment]
A side-note about this hdf5 version mismatch: CDO in fact depends on hdf5 it *two* different ways

* there is one operator (_import_cmsaf_) which reads hdf5 files from satellites -> direct dependency
* the current netcdf-4.x is build on top of hdf5 -> indirect dependency

Of course at runtime there can only be *one* hdf5 library used, that's why building CDO against different hdf5 libs can lead to ... unexpected behavior. 

The second issue is _compile time_ vs. _runtime_: If you do not inject correct ```rpaths``` during the building, and use nasty stuff like ```LD_LIBRARY_PATH``` instead, the runtime version of hdf5 might be different to the  compile-time version.
